cluster mempool: cluster linearization algorithm

sipa opened the pull request:
Part of cluster mempool: #30289 


This introduces low-level cluster linearization code, including tests and some benchmarks. It is currently not hooked up to anything.

Ultimately, what this PR adds is a function `Linearize` which operates on instances of `DepGraph` (instances of which represent pre-processed transaction clusters) to produce and/or improve linearizations for that cluster.

To provide assurance, the code heavily relies on fuzz tests. A novel approach is used here, where the fuzz input is parsed using the serialization.h framework rather than `FuzzedDataProvider`, with a custom serializer/deserializer for `DepGraph` objects. By including serialization, it's possible to ascertain that the format can represent every relevant cluster, as well as potentially permitting the construction of ad-hoc fuzz inputs from clusters (not included in this PR, but used during development).

---

The `Linearize(depgraph, iteration_limit, rng_seed, old_linearization)` function is an implementation of the (single) LIMO algorithm, with the $S$ in every iteration found as the best out of (a) the best remaining ancestor set and (b) randomized computationally-bounded search. It incrementally builds up a linearization by finding good topologically-valid subsets to move to the front, in such a way that the resulting linearization has a diagram that is at least as good as the `old_linearization` passed in (if any).
* Despite using both best ancestor set and search, this is not Double LIMO, as no intersections between these are involved; just the best of the two.
* The `iteration_limit` and `rng_seed` only control the (b) randomized search. Even with 0 iterations, the result will be as good as the old linearization, and the included sets at every point will have a feerate at least as high as the best remaining ancestor set at that point.

The search algorithm used in the (b) step is very basic, and largely matches Section 2.1 of [How to Linearize your Cluster.](https://delvingbitcoin.org/t/how-to-linearize-your-cluster/303#h-21-searching-6). See #30286 for optimizations to make it more efficient.

For background and references, see [Introduction to cluster linearization](https://delvingbitcoin.org/t/introduction-to-cluster-linearization/1032).

sipa committed:
clusterlin: introduce cluster_linearize.h with Cluster and DepGraph types

This primarily adds the DepGraph class, which encapsulates precomputed
ancestor/descendant information for a given transaction cluster, with a
number of utility features (inspectors for set feerates, computing
reduced parents/children, adding transactions, adding dependencies), which
will become needed in future commits.

sipa committed:
clusterlin: add Linearize function

This adds a first version of the overall linearization interface, which given
a DepGraph constructs a good linearization, by incrementally including good
candidate sets (found using AncestorCandidateFinder and SearchCandidateFinder).

sipa committed:
clusterlin: add SearchCandidateFinder class

Similar to AncestorCandidateFinder, this encapsulates the state needed for
finding good candidate sets using a search algorithm.

sipa committed:
clusterlin: add AncestorCandidateFinder class

This is a class that encapsulates precomputed ancestor set feerates, and
presents an interface for getting the best remaining ancestor set.

sipa committed:
tests: framework for testing DepGraph class

This introduces a bespoke fuzzing-focused serialization format for DepGraphs,
and then tests that this format can represent any graph, roundtrips, and then
uses that to test the correctness of DepGraph itself.

This forms the basis for future fuzz tests that need to work with interesting
graphs.

sipa committed:
clusterlin: permit passing in existing linearization to Linearize

This implements the LIMO algorithm for linearizing by improving an existing
linearization. See
https://delvingbitcoin.org/t/limo-combining-the-best-parts-of-linearization-search-and-merging
for details.

sipa committed:
clusterlin: use bounded BFS exploration (optimization)

Switch to BFS exploration of the search tree in SearchCandidateFinder
instead of DFS exploration. This appears to behave better for real
world clusters.

As BFS has the downside of needing far larger search queues, switch
back to DFS temporarily when the queue grows too large.

sipa committed:
clusterlin: randomize the SearchCandidateFinder search order

To make search non-deterministic, change the BFS logic from always picking
the first queue item to randomly picking the first or second queue item.

sipa committed:
bench: Candidate finding and linearization benchmarks

Add benchmarks for known bad graphs for the purpose of search (as
an upper bound on work per search iterations) and ancestor sorting
(as an upper bound on linearization work with no search iterations).

theuni commented on src/util/bitset.h:
```diff
@@ -0,0 +1,471 @@
+#include <type_traits>
+
+#ifdef _MSC_VER
+#  include <intrin.h>
```
Why is this needed?

sr-gi commented:
Some more benchmarks.

MacBook M3 Max:

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            1,678.30 |          595,841.56 |    0.4% |      1.11 | `LinearizeNoIters16TxWorstCase`
|            4,975.16 |          200,998.49 |    0.1% |      1.10 | `LinearizeNoIters32TxWorstCase`
|           10,624.85 |           94,118.94 |    0.1% |      1.10 | `LinearizeNoIters48TxWorstCase`
|           18,570.63 |           53,848.46 |    0.1% |      1.10 | `LinearizeNoIters64TxWorstCase`
|           38,962.09 |           25,665.97 |    0.1% |      1.10 | `LinearizeNoIters75TxWorstCase`
|           65,933.96 |           15,166.69 |    0.1% |      1.06 | `LinearizeNoIters99TxWorstCase`

|            ns/iters |             iters/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               37.31 |       26,803,657.99 |    0.1% |      1.10 | `LinearizePerIter16TxWorstCase`
|               26.62 |       37,561,209.12 |    0.2% |      1.10 | `LinearizePerIter32TxWorstCase`
|               25.79 |       38,768,485.22 |    0.2% |      1.10 | `LinearizePerIter48TxWorstCase`
|               26.05 |       38,392,847.23 |    0.1% |      1.10 | `LinearizePerIter64TxWorstCase`
|               37.99 |       26,322,699.89 |    0.2% |      1.09 | `LinearizePerIter75TxWorstCase`
|               38.21 |       26,169,082.76 |    0.1% |      1.10 | `LinearizePerIter99TxWorstCase`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              375.76 |        2,661,273.20 |    0.3% |      1.11 | `PostLinearize16TxWorstCase`
|            1,295.00 |          772,200.32 |    0.1% |      1.10 | `PostLinearize32TxWorstCase`
|            2,850.98 |          350,756.38 |    0.0% |      1.10 | `PostLinearize48TxWorstCase`
|            5,155.30 |          193,975.14 |    0.0% |      1.10 | `PostLinearize64TxWorstCase`
|            6,804.84 |          146,954.16 |    0.1% |      1.06 | `PostLinearize75TxWorstCase`
|           11,696.22 |           85,497.68 |    0.0% |      1.10 | `PostLinearize99TxWorstCase`

Raspberry Pi 4 (arm32):

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|----------:|:----------
|           24,979.12 |           40,033.43 |    0.1% |       62,303.01 |       37,432.31 |  1.664 |      1.10 | `LinearizeNoIters16TxWorstCase`
|           79,796.77 |           12,531.84 |    0.0% |      199,042.69 |      119,579.10 |  1.665 |      1.08 | `LinearizeNoIters32TxWorstCase`
|          206,049.75 |            4,853.20 |    0.0% |      514,152.74 |      308,732.11 |  1.665 |      1.10 | `LinearizeNoIters48TxWorstCase`
|          339,314.40 |            2,947.12 |    0.0% |      866,686.12 |      508,414.12 |  1.705 |      1.10 | `LinearizeNoIters64TxWorstCase`
|          574,288.94 |            1,741.28 |    0.0% |    1,309,819.21 |      860,366.98 |  1.522 |      1.10 | `LinearizeNoIters75TxWorstCase`
|          983,808.56 |            1,016.46 |    0.0% |    2,278,204.68 |    1,474,149.12 |  1.545 |      1.10 | `LinearizeNoIters99TxWorstCase`

|            ns/iters |             iters/s |    err% |       ins/iters |       cyc/iters |    IPC |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|----------:|:----------
|              362.94 |        2,755,255.49 |    0.1% |          862.03 |          543.89 |  1.585 |      1.10 | `LinearizePerIter16TxWorstCase`
|              290.11 |        3,446,974.76 |    0.2% |          692.02 |          434.73 |  1.592 |      1.10 | `LinearizePerIter32TxWorstCase`
|              529.35 |        1,889,094.80 |    0.1% |        1,319.02 |          793.26 |  1.663 |      1.11 | `LinearizePerIter48TxWorstCase`
|              530.00 |        1,886,788.68 |    0.2% |        1,324.44 |          794.26 |  1.668 |      1.11 | `LinearizePerIter64TxWorstCase`
|              782.35 |        1,278,197.39 |    0.5% |        2,079.81 |        1,172.37 |  1.774 |      1.06 | `LinearizePerIter75TxWorstCase`
|              737.95 |        1,355,112.46 |    0.5% |        1,684.40 |        1,105.59 |  1.524 |      1.07 | `LinearizePerIter99TxWorstCase`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|----------:|:----------
|            5,010.36 |          199,586.29 |    0.8% |       10,606.00 |        7,506.09 |  1.413 |      1.10 | `PostLinearize16TxWorstCase`
|           14,365.91 |           69,609.26 |    0.0% |       34,330.00 |       21,530.90 |  1.594 |      1.09 | `PostLinearize32TxWorstCase`
|           32,751.91 |           30,532.57 |    0.0% |       76,671.01 |       49,085.98 |  1.562 |      1.10 | `PostLinearize48TxWorstCase`
|           55,794.83 |           17,922.81 |    0.0% |      131,423.02 |       83,621.11 |  1.572 |      1.10 | `PostLinearize64TxWorstCase`
|           81,534.00 |           12,264.82 |    1.6% |      210,784.02 |      120,341.56 |  1.752 |      1.09 | `PostLinearize75TxWorstCase`
|          144,665.53 |            6,912.50 |    0.0% |      399,573.04 |      216,805.74 |  1.843 |      1.10 | `PostLinearize99TxWorstCase`

Raspberry Pi 4 (aarch64):

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|----------:|:----------
|           13,765.93 |           72,643.09 |    0.1% |       37,001.00 |       20,625.64 |  1.794 |     11.01 | `LinearizeNoIters16TxWorstCase`
|           41,091.94 |           24,335.67 |    0.2% |      107,447.00 |       61,567.66 |  1.745 |     10.99 | `LinearizeNoIters32TxWorstCase`
|           73,005.68 |           13,697.56 |    0.1% |      210,986.00 |      109,380.97 |  1.929 |     10.67 | `LinearizeNoIters48TxWorstCase`
|          118,418.69 |            8,444.61 |    0.0% |      348,914.00 |      177,423.52 |  1.967 |     11.01 | `LinearizeNoIters64TxWorstCase`
|          229,345.84 |            4,360.23 |    0.0% |      618,749.67 |      343,589.10 |  1.801 |     11.01 | `LinearizeNoIters75TxWorstCase`
|          377,809.11 |            2,646.84 |    0.0% |    1,021,644.01 |      565,996.91 |  1.805 |     11.00 | `LinearizeNoIters99TxWorstCase`

|            ns/iters |             iters/s |    err% |       ins/iters |       cyc/iters |    IPC |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|----------:|:----------
|              231.99 |        4,310,566.43 |    0.0% |          643.14 |          347.59 |  1.850 |     11.00 | `LinearizePerIter16TxWorstCase`
|              190.47 |        5,250,207.80 |    0.1% |          525.81 |          285.35 |  1.843 |     11.03 | `LinearizePerIter32TxWorstCase`
|              175.35 |        5,702,969.82 |    0.1% |          492.10 |          262.75 |  1.873 |     11.01 | `LinearizePerIter48TxWorstCase`
|              175.70 |        5,691,481.30 |    0.1% |          494.56 |          263.25 |  1.879 |     11.00 | `LinearizePerIter64TxWorstCase`
|              293.22 |        3,410,384.42 |    0.2% |          763.38 |          439.30 |  1.738 |     11.01 | `LinearizePerIter75TxWorstCase`
|              293.55 |        3,406,519.94 |    0.1% |          766.88 |          439.77 |  1.744 |     11.01 | `LinearizePerIter99TxWorstCase`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|----------:|:----------
|            2,915.23 |          343,026.59 |    0.1% |        5,325.00 |        4,368.47 |  1.219 |     11.01 | `PostLinearize16TxWorstCase`
|            9,830.26 |          101,726.71 |    0.0% |       17,247.00 |       14,729.95 |  1.171 |     11.00 | `PostLinearize32TxWorstCase`
|           20,550.14 |           48,661.46 |    0.0% |       35,742.00 |       30,794.09 |  1.161 |     11.00 | `PostLinearize48TxWorstCase`
|           35,211.73 |           28,399.62 |    0.0% |       61,199.00 |       52,763.93 |  1.160 |     11.00 | `PostLinearize64TxWorstCase`
|           31,801.73 |           31,444.83 |    0.0% |       95,378.00 |       47,653.77 |  2.001 |     11.00 | `PostLinearize75TxWorstCase`
|           52,955.80 |           18,883.67 |    0.0% |      162,443.00 |       79,349.66 |  2.047 |     11.00 | `PostLinearize99TxWorstCase`

instagibbs commented:
> Ancestor-set based presplitting inside the search is replaced with using the best ancestor set as one of the LIMO set choices.

Can this be elaborated a bit? I've taken a few minutes looking at the commits and it wasn't immediately clear how this all maps to existing public discussions and the comments:

```
        // This is an implementation of the (single) LIMO algorithm:
        // https://delvingbitcoin.org/t/limo-combining-the-best-parts-of-linearization-search-and-merging/825
        // where S is instantiated to be the result of a bounded search, which itself is seeded
        // with the best prefix of what remains of the input linearization, or the best ancestor set.
```
thanks!

instagibbs reviewed (commented):
glanced in the context of checking what of VecDeque/BitSet actually required

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -0,0 +1,390 @@
+    DepGraph depgraph(cluster);
+    SanityCheck(depgraph);
+
+    // Verify that ancestry is computed correctly.
```
for assurance shouldn't we also be checking the descendants are being computed as expected? Or is that covered somewhere else?

instagibbs reviewed (commented):
couldn't find any logic issues in the parts I complete understood, I've run the fuzzer on each commit and reviewed through  https://github.com/bitcoin/bitcoin/pull/30126/commits/9633b40722fd9295b93baaf9914b31b9dec96f45

a couple general comments:

1) The bespoke DepGraphFormatter is pretty artisanal and as a reviewer I'm relying heavily on the cluster->depgraph->check depgraph matches -> serialization round-trip for asserting correctness. Unit tests-as-documentation is one thing I'm kind of missing, especially with the more involved parts.
2) Matching the actual LIMO commit with the LIMO literature/theory available is difficult for me(e.g., what's sufficient vs necessary). I know there's a new writeup coming, so I'll wait for that before investigating further.

sdaftuar commented on src/cluster_linearize.h:
```diff
@@ -5,6 +5,7 @@
 #ifndef BITCOIN_CLUSTER_LINEARIZE_H
```
nit: if you touch up this commit, I think the commit message has a couple typos:

`This is a class that encapsulated precomputes ancestor set feerates, and`

should read "encapsulates precomputed ancestor set feerates"

sdaftuar reviewed (commented):
Just reviewed through the first 3 commits so far.

glozow reviewed (commented):
Looked at the first 4 commits so far, just doing code review and trying the algos out in `MiniMiner`.

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -235,6 +235,106 @@ struct DepGraphFormatter
+                }
+            }
+        }
+        return {std::move(best), max_iterations - iterations_left};
```
I'm assuming that it's only used for stronger invariants checks in the fuzz/test harness

sipa committed:
clusterlin: add chunking algorithm

A fuzz test is added which verifies various of its expected properties, including
correctness

sipa committed:
clusterlin: add LinearizationChunking class

It encapsulates a given linearization in chunked form, permitting arbitrary
subsets of transactions to be removed from the linearization. Its purpose
is adding the Intersect function, which is a crucial operation that will
be used in a further commit to make Linearize improve existing linearizations.

sipa committed:
bench: add cluster linearization improvement benchmark

ismaelsadeeq commented on src/cluster_linearize.h:
```diff
@@ -0,0 +1,167 @@
+     **/
+    void AddDependency(ClusterIndex parent, ClusterIndex child) noexcept
+    {
+        // To each ancestor of the parent, add as descendants the descendants of the child.
```
maybe just return early in that case and avoid the looping?

ismaelsadeeq commented on src/cluster_linearize.h:
```diff
@@ -0,0 +1,168 @@
+     **/
+    void AddDependency(ClusterIndex parent, ClusterIndex child) noexcept
+    {
+        // To each ancestor of the parent, add as descendants the descendants of the child.
```
Should we return early if dependency already exist ?

achow101 commented:
c192b30156ae41638291010b40b874479ea1943c "clusterlin: add algorithms for connectedness/connected components" seems to be only relevant for the optimizations in #30286, so could be dropped from here?

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -320,6 +351,7 @@ FUZZ_TARGET(clusterlin_chunking)
     try {
         reader >> Using<DepGraphFormatter>(depgraph);
     } catch (const std::ios_base::failure&) {}
+    MakeConnected(depgraph);
```
Do we always want to make the graphs connected? Seems like a possible reduction in coverage?

instagibbs replied:
I'm not particularly bothered by it for the reasons you mentioned. I'll let others weigh in.

instagibbs reviewed (commented):
another pass, through https://github.com/bitcoin/bitcoin/pull/30126/commits/23496cb4d43327c3d27401a531586bcfc985b6ad

changes since last look make sense, I would like to spend more time validating the optimizations

ismaelsadeeq reviewed (commented):
Reviewed:

- [x] 4409a282d7fe7a0ebee67c8fb9fb4ef157ed5883
- [x] f5fa49f589f477caa5a4ca41d8331acdca6d7298
- [x] 48f086eb624a5371b846e908b4702b28726465fa
- [x] ccde33054e9573bc9c263acd9746b8003578cb29
- [x] 6fdb3d07d4350b0d93bda5765ea1053c29866c0a
- [x] 448b8b977920b8d2f90008fa1939d3d3bba63668
- [x] 896b52337be5765cb3a7d3f0458475f2f8e2cb3a

I spent time understanding the [cluster linearization theory](https://delvingbitcoin.org/t/introduction-to-cluster-linearization/1032).
The serialization and de-serialization code in f5fa49f589f477caa5a4ca41d8331acdca6d7298 are well documented 💯

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -253,6 +253,109 @@ std::vector<FeeFrac> ChunkLinearization(const DepGraph<SetType>& depgraph, Span<
+                if (accumulator.transactions == subset.transactions) break;
+                // Otherwise update the accumulator feerate.
+                accumulator.feerate += m_depgraph.FeeRate(to_add);
+                // If that does result in something better, return that.
```
> returning subset in this case would not be correct, as it's possible that a bigger intersection still has a higher feerate.

Ah I had forgotten this fact, makes total sense because it doesn't hurt to just stop here.

instagibbs commented on src/test/util/cluster_linearize.h:
```diff
@@ -0,0 +1,255 @@
+                ++insert_distance;
+            }
+            rebuilt_order.insert(rebuilt_order.end() - insert_distance, idx);
+            s << VARINT(diff + insert_distance);
```
Don't go out of your way for now, this may just be a mental block for myself only. I worked out some examples myself and it works, I just can't mentally model the skipping somehow.

instagibbs replied:
yes this was very insightful, thank you

instagibbs commented on src/test/util/cluster_linearize.h:
```diff
@@ -0,0 +1,255 @@
+            ClusterIndex topo_idx = reordering[idx];
+            depgraph.AddTransaction(topo_depgraph.FeeRate(topo_idx));
+        }
+        for (ClusterIndex idx = 0; idx < reordering.size(); ++idx) {
```
oh whoops, wrong one. I meant for the loop above!

instagibbs commented:
reviewed through https://github.com/bitcoin/bitcoin/pull/30126/commits/2003bb8a279c8891e55bab190ca36f0c6c8697ea

via `git range-diff master 23496cb 2003bb8a279c8891e55bab190ca36f0c6c8697ea`

instagibbs commented on src/bench/cluster_linearize.cpp:
```diff
@@ -112,6 +127,23 @@ void BenchLinearizeNoItersWorstCase(ClusterIndex ntx, benchmark::Bench& bench)
+ * iterations, similar to the previous function.
+ */
+template<typename SetType>
+void BenchLinearizeNoItersWorstCaseLIMO(ClusterIndex ntx, benchmark::Bench& bench)
```
Why is this benchmark "LIMO" but not `BenchLinearizeNoItersWorstCaseAnc`?

instagibbs replied:
ah, parsing the name wrong basically, thanks

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -323,9 +331,21 @@ class LinearizationChunking
-        // Rechunk what remains of m_linearization.
-        m_chunks.clear();
-        BuildChunks();
+        if (GetChunk(0).transactions == subset) {
```
re: `clusterlin: optimize rechunking in LinearizationChunking`

benchmark doesn't seem to show much improvement for this change. I understood this as purely an efficiency improvement which should be reflected in bench results?

instagibbs commented:
ACK https://github.com/bitcoin/bitcoin/pull/30126/commits/6160ccf9b4327649e9bb0293fba630a10b3befc3

reviewed via `git range-diff master 2003bb8a279c8891e55bab190ca36f0c6c8697ea 6160ccf9b4327649e9bb0293fba630a10b3befc3`

I didn't validate that the BFS optimisation for search improved results, but it at least doesn't seem to hurt benchmark performance.

glozow reviewed (commented):
code review ACK, planning to spend more time reviewing the tests

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/30126/commits/cad318fa843f411e52c6761a4882bfaf0ad21812

via `git range-diff master 6160ccf9b4327649e9bb0293fba630a10b3befc3 cad318fa843f411e52c6761a4882bfaf0ad21812`

glozow reviewed (commented):
ACK cad318fa843

I did code review to convince myself that the code does what it's supposed to be doing / what's described in the delving post. Everything is a bit abstract so I manually tested by writing some naive versions of `BlockAssembler` and  `MiniMiner` using cluster_linearize. I also code reviewed the tests and introduced bugs to check their coverage. I didn't really dive into `DepGraphFormatter` but it seems to work.

glozow commented on src/bench/cluster_linearize.cpp:
```diff
@@ -0,0 +1,139 @@
+    return depgraph;
+}
+
+// Construct a difficult graph. These need at least sqrt(2^(n-1)) iterations in the best
+// known algorithms (purely empirically determined).
+template<typename SetType>
+DepGraph<SetType> MakeHardGraph(ClusterIndex ntx)
```
Matches my diagrams. Thanks for the additional context! :+1:

glozow replied:
Ah I didn't notice this yesterday - I think you may have swapped odd and even? Same in the commented mermaid diagram code just added.

glozow replied:
Hm, I was confused because when I was looking at the code, I expected the branch with `depgraph.AddDependency(i, 4)` to have lots of T4-->Tx in its diagram. And the branch with `depgraph.AddDependency(i, 3)` to have lots of T3-->Tx.

{T0...T9} is 10 transactions, and {T0...T10} is 11 transactions. Am I missing something?

glozow replied:
Whew, so not wholly unmeritorious.

sdaftuar reviewed (commented):
Code review ACK cad318fa843f411e52c6761a4882bfaf0ad21812.  Left some non-blocking nits/suggestions.

I've run the fuzz tests in earlier versions of this PR for a while, but I haven't done so with the latest branch -- will do that and also post some benchmarks in a bit.

sdaftuar commented:
Benchmark results on my ryzen 7995wx:

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            1,474.01 |          678,422.17 |    0.3% |      0.01 | `LinearizeNoIters16TxWorstCaseAnc`
|            1,473.48 |          678,667.50 |    0.2% |      0.01 | `LinearizeNoIters16TxWorstCaseLIMO`
|            4,763.02 |          209,951.04 |    0.1% |      0.01 | `LinearizeNoIters32TxWorstCaseAnc`
|            5,138.13 |          194,623.26 |    0.0% |      0.01 | `LinearizeNoIters32TxWorstCaseLIMO`
|           10,296.19 |           97,123.33 |    0.0% |      0.01 | `LinearizeNoIters48TxWorstCaseAnc`
|           11,148.56 |           89,697.71 |    0.0% |      0.01 | `LinearizeNoIters48TxWorstCaseLIMO`
|           17,321.56 |           57,731.51 |    0.1% |      0.01 | `LinearizeNoIters64TxWorstCaseAnc`
|           19,224.32 |           52,017.44 |    0.0% |      0.01 | `LinearizeNoIters64TxWorstCaseLIMO`
|           25,346.32 |           39,453.45 |    0.1% |      0.01 | `LinearizeNoIters75TxWorstCaseAnc`
|           28,625.57 |           34,933.81 |    0.1% |      0.01 | `LinearizeNoIters75TxWorstCaseLIMO`
|           43,628.14 |           22,920.98 |    0.1% |      0.01 | `LinearizeNoIters99TxWorstCaseAnc`
|           49,273.81 |           20,294.76 |    0.1% |      0.01 | `LinearizeNoIters99TxWorstCaseLIMO`

|            ns/iters |             iters/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               13.81 |       72,390,978.12 |    0.1% |      0.01 | `LinearizePerIter16TxWorstCase`
|                9.62 |      103,944,701.42 |    0.2% |      0.01 | `LinearizePerIter32TxWorstCase`
|                9.29 |      107,658,429.17 |    0.2% |      0.01 | `LinearizePerIter48TxWorstCase`
|                9.30 |      107,518,263.44 |    0.2% |      0.01 | `LinearizePerIter64TxWorstCase`
|               10.21 |       97,957,584.37 |    0.1% |      0.01 | `LinearizePerIter75TxWorstCase`
|               10.07 |       99,325,875.28 |    0.2% |      0.01 | `LinearizePerIter99TxWorstCase`

glozow reviewed (commented):
reACK df749c06e9d via range-diff

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/30126/commits/df749c06e9d45ba92a4699e1d3e58522f93ad6bb

via `range-diff master cad318fa843f411e52c6761a4882bfaf0ad21812 df749c06e9d45ba92a4699e1d3e58522f93ad6bb`

Didn't verify the mermaid diagrams in the comments.

At this point could we maybe not rebase on master unless necessary? :)

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/30126/commits/448374f2c901c48824904357b4e1297105e97315 

via `git range-diff master df749c06e9d45ba92a4699e1d3e58522f93ad6bb 448374f2c901c48824904357b4e1297105e97315`

only changes to MakeHardGraph to reflect the intended topology

glozow commented:
reACK 448374f2c90 via range-diff

instagibbs commented:
reACK 647fa37cdbadbeebba147ca6b24e138559cffaaf

via `git range-diff master 448374f2c901c48824904357b4e1297105e97315 647fa37cdbadbeebba147ca6b24e138559cffaaf`

I traced through the generated diagrams + code, seems to match. Didn't verify difficuly.

sdaftuar commented:
ACK 647fa37cdbadbeebba147ca6b24e138559cffaaf

In addition to some manual testing, I ran all the fuzz targets on my machine for a while as well.

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/290.

ismaelsadeeq commented:
post merge ACK 647fa37cdbadbeebba147ca6b24e138559cffaaf
